align_system/utils/incontext_utils.py

```diff
@@ -0,0 +1,287 @@
+        
+        if icl_strategy == "random":
+            selected_icl_examples = random.sample(possible_icl_examples, n_icl_examples)
+        elif icl_strategy == "bert_similarity":
+            possible_icl_prompts = [icl_sample["prompt"] for icl_sample in possible_icl_examples]
+            # Create similarity scores between the ICL samples and find top-k indices
+            from bert_score import score
+            _, _, F1 = score([prompt_to_match]*len(possible_icl_prompts), possible_icl_prompts, lang="en")
+            _, indices = torch.topk(F1, n_icl_examples)
+
+            selected_icl_examples = [possible_icl_examples[i] for i in indices]
+        else:
+            raise ValueError(f'"{icl_strategy}" is not a valid incontext method. Please use "random" or '
+                                '"bert_similarity"')
+
+        return selected_icl_examples
+
+
+class ComparativeRegressionIncontextExampleGenerator(IncontextExampleGenerator):
+    def __init__(self,
+                 incontext_settings, 
+                 target_kdmas, 
+                 log,
+                 **kwargs):
+        self.incontext_settings = incontext_settings
+        self.target_kdmas = target_kdmas
+        self.log = log
+        self.set_icl_datasets()
+    
+    def set_icl_datasets(self):
+        icl_datasets = {}
+        incontext_data = self.read_icl_dataset_files()
+        
+        # Add each target to icl_datasets
+        for target_kdma in self.target_kdmas:
+            sys_kdma_name = target_kdma['kdma']
+            icl_datasets[sys_kdma_name] = []
+            kdma_incontext_data = incontext_data[sys_kdma_name]
+            
+            # Add each examples to icl_datasets
+            for example in kdma_incontext_data:
+                
+                # Get example prompt
+                character_info = outlines_prompts_utils.get_relevant_structured_character_info(example['state'].characters)
+                icl_scenario_description = scenario_state_description_with_relevant_char_info(example['state'], character_info)
+                icl_choices_with_outcomes = {}
+                for choice in example['choices']:
+                    # TODO: Include outcome prediction for ICL examples?
+                    icl_choices_with_outcomes[choice] = {'predicted_outcome':None}
+                icl_prompt = comparative_kdma_score_prediction_prompt(icl_scenario_description,
+                                                                    icl_choices_with_outcomes,
+                                                                    sys_kdma_name)
+                # Get example response
+                icl_response = {}
+                for action, choice, kdma_value in zip(example['actions'], example['choices'], example["kdma_values"]):
+                    # Only include choice if there is a ground truth KDMA value available
+                    if kdma_value is None:
+                        continue
```

Comment:
This skips non-labeled actions in the response schema, but they are still included in the prompt. This matches the DRE behavior, but we may want to considering pre-filtering before the prompt is generated

Reply:
Good point, fixed here: https://github.com/ITM-Kitware/align-system/pull/115/commits/c5c7c851ff940c4dbee37272a390df8b7b1e000c